smolyay/samples.py

```diff
@@ -0,0 +1,362 @@
+        and other parameters.
+        """
+        pass
+
+
+class NestedUnidimensionalPointSet(UnidimensionalPointSet):
+    """Set of unidimensional points with levels
+
+    A set of unique unidimensional points to be used as the sampling points
+    in a tensor product grid or sparse grid. These unidimensional points are
+    associated with certain polynomial families, and the combination of
+    unique points and polynomial families are the basis for different types
+    of quadrature rules that are used in numerical integration and in
+    approximation.
+
+    The number of unique points is controlled by `num_points`, which describes
+    the amount of points added by each level.
+    `points` describes the unique points added as the number of levels increase,
+    and the order of elements is reflective of the level each element first
+    appears in. Some point set are nested, meaning that the points in a previous
+    level reappear in following levels. However, `points` does not attempt to
+    capture this behavior and should be free of duplicates.
+
+    Parameters
+    ----------
+    max_level ; int
+        the maximum level the points are used for
+    """
+
+    def __init__(self, max_level):
+        super().__init__()
+        self.max_level = max_level
+        self._num_points = None
+        self._start_level = None
+        self._end_level = None
+
+    @property
+    def max_level(self):
+        """int: maximum level to compute points for."""
+        return self._max_level
+
+    @max_level.setter
+    def max_level(self, value):
+        self._max_level = value
+        self._valid_cache = True
+
+    @property
+    def num_points(self):
+        """list: number of points per level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._num_points
+
+    @property
+    def start_level(self):
+        """list: the starting index of each level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._start_level
+
+    @property
+    def end_level(self):
+        """list: the ending index of each level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._end_level
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+
+    Parameters
+    ----------
+    degree : int
+        degree of the Chebyshev polynomial of the first kind to get extrema from
+    """
+
+    def __init__(self, degree):
```

Comment:
Add domain to constructor. You can pick what order makes sense (domain first or not), then be consistent in all children